refactor: Setup typescript project references across workflow, core, and cli

netroy opened the pull request:
This makes navigating between packages much better. After this change, if you ctrl+click on an import from another package, you will go to the source file, and not the `dist/*.d.ts` file.
Also, while developing on `cli`, you don't need to keep building `core` and `workflow`.

as a bonus, most of of the cyclic dependencies have been removed.

netroy committed:
refactor: use consistent folder structure across workflow, core, and cli

netroy committed:
setup typescript project references across workflow, core, and cli

ivov commented on packages/cli/src/PublicApi/types.d.ts:
```diff
@@ -2,13 +2,13 @@
+import type { WorkflowEntity } from '~/databases/entities/WorkflowEntity';
 
-import * as UserManagementMailer from '../UserManagement/email/UserManagementMailer';
+import * as UserManagementMailer from '~/UserManagementemail/UserManagementMailer';
```
```suggestion
import * as UserManagementMailer from '~/UserManagement/email/UserManagementMailer';
```

ivov replied:
Surprising that VSCode unable to flag broken imports.

ivov reviewed (dismissed):
`Go to definition` confirmed working.

1. From a clean install, I get this issue on `build`

```
n8n:build: > n8n@0.201.0 build
n8n:build: > node scripts/build.mjs
n8n:build:
n8n:build: Typescript Compilation failed:
n8n:build: src/CurlConverterHelper.ts(2,27): error TS1261: Already included file name '/Users/ivov/Development/n8n/packages/cli/src/curlconverter.d.ts' differs from file name '/Users/ivov/Development/n8n/packages/cli/src/CurlConverter.d.ts' only in casing.
n8n:build:   The file is in the program because:
n8n:build:     Imported via 'curlconverter' from file '/Users/ivov/Development/n8n/packages/cli/src/CurlConverterHelper.ts'
n8n:build:     Matched by include pattern 'src/**/*.ts' in '/Users/ivov/Development/n8n/packages/cli/tsconfig.json'
n8n:build:
n8n:build: npm ERR! Lifecycle script `build` failed with error:
n8n:build: npm ERR! Error: command failed
```

2. If I ts-ignore the above issue, `build` finishes and I get this issue on `start`

<details>
  <summary>Issue</summary>

```
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/Logger'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/BaseCommand.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(Use `node --trace-warnings ...` to show where the warning was created)
(node:29704) TypeError Plugin: n8n: Cannot read properties of undefined (reading 'join')
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/Logger'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/executeBatch.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/config'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/start.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/config'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/webhook.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) TypeError Plugin: n8n: Cannot read properties of undefined (reading 'join')
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/Logger'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/db/revert.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/Logger'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/export/credentials.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/Logger'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/export/workflow.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/Logger'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/import/credentials.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/Logger'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/import/workflow.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/Db'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/list/workflow.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/Db'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/update/workflow.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
(node:29704) [MODULE_NOT_FOUND] Error Plugin: n8n: Cannot find module '~/Db'
Require stack:
- /Users/ivov/Development/n8n/packages/cli/src/commands/user-management/reset.ts
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
- /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
- /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
- /Users/ivov/Development/n8n/packages/cli/bin/n8n
module: @oclif/config@1.18.5
task: toCached
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
See more details with DEBUG=*
 ›   Error: command start not found
```
</details>

3. Failing tests in `cli` on sqlite 
```
Test Suites: 4 failed, 19 passed, 23 total
Tests:       26 failed, 1 skipped, 295 passed, 322 total
Snapshots:   0 total
Time:        18.928 s, estimated 28 s
Ran all test suites.
```

4. We should aim for consistency in wildcard imports vs named imports, e.g. `import * as ActiveExecutions from '~/ActiveExecutions'` vs `import { ActiveExecutions } from '~/ActiveExecutions'`. I'd personally prefer named imports.

5. We should aim for consistency in index imports. `import { X } from '~/index'` vs `import { X } from '..'`. I'd personally prefer the more explicit variant.

6. We should aim for consistency in relative paths. `import config from './config'` vs `import config from '~/config'`. I'd personally prefer package-relative paths.

7. Should we lint to enforce that all relative paths are package-relative paths instead of regular relative paths? Can come later.

8. Some more path aliases might be useful, e.g. `~/databases/entities` → `@models`. Can come later.

netroy commented:
- [DB tests pass](https://github.com/n8n-io/n8n/actions/runs/3397211474/jobs/5649132810).
- [Docker image works](https://hub.docker.com/layers/n8nio/n8n/N8N-5129-N8N-5130-N8N-4442/images/sha256-4dda61a611379cf8ac5060755ce6dd0877b1c6b0e6197441b30774bc7218edfd?context=explore)

>     1. From a clean install, I get this issue on `build`
>     2. If I ts-ignore the above issue, `build` finishes and I get this issue on `start`
>     3. Failing tests in `cli` on sqlite

Can you please try these again?



>     4. We should aim for consistency in wildcard imports vs named imports, e.g. `import * as ActiveExecutions from '~/ActiveExecutions'` vs `import { ActiveExecutions } from '~/ActiveExecutions'`. I'd personally prefer named imports.
I personally prefer named imports as well, but perhaps that can be another PR?



>     5. We should aim for consistency in index imports. `import { X } from '~/index'` vs `import { X } from '..'`. I'd personally prefer the more explicit variant.
I'd prefer that we never use `~/index`. it's the largest culprit behind cyclic dependencies.
I deleted that file, and make the `cli` compile, test, and run without it. I've added the file back for backward compatibility. But, it's not used anymore.



>     6. We should aim for consistency in relative paths. `import config from './config'` vs `import config from '~/config'`. I'd personally prefer package-relative paths.
Agreed, and done. I also changed `~/` to `@/`, to be consistent with what the frontend code is using.



>     7. Should we lint to enforce that all relative paths are package-relative paths instead of regular relative paths? Can come later.
Might be worth trying, but only if we see people not using package-relative imports.



>     8. Some more path aliases might be useful, e.g. `~/databases/entities` → `@models`. Can come later.
Added `@db/` for `src/database/` as a first example. We could add more like `@um/` for `src//UserManagement/` in another PR.

netroy committed:
stop using `__dirname, '..', '..', '..'`

it's very confusing, and breaks when the file path changes

ivov reviewed (commented):
Seeing multiple type issues from cross-package imports. From a clean install and after restarting TS language server.
![image](https://user-images.githubusercontent.com/44588767/200281118-27dc9de7-da2a-4bb0-85f0-64e5960943b4.png)

> I also changed `~/` to `@/`, to be consistent with what the frontend code is using.

Matter of preference but I find it better to keep `@` for path shortcuts and `~` for package root, because `~` is clear in meaning and because using `@` for both package root and for shortcuts can lead to confusion: `@/someDir` vs `@someDir`. We can also change FE after discussion.

netroy commented:
> Seeing multiple type issues from cross-package imports. From a clean install and after restarting TS language server. ![image](https://user-images.githubusercontent.com/44588767/200281118-27dc9de7-da2a-4bb0-85f0-64e5960943b4.png)

looks like an issue with eslint https://github.com/typescript-eslint/typescript-eslint/issues/2094 . Let me see if I can find a workaround

> > I also changed `~/` to `@/`, to be consistent with what the frontend code is using.
> 
> Matter of preference but I find it better to keep `@` for path shortcuts and `~` for package root, because `~` is clear in meaning and because using `@` for both package root and for shortcuts can lead to confusion: `@/someDir` vs `@someDir`. We can also change FE after discussion.

I'd prefer to make that another PR where we change the FE and BE code together.

netroy commented:
Looks like the eslint errors aren't due to project references, but because of the package-relative imports (or at-least how they are setup).

netroy committed:
Merge remote-tracking branch 'origin/master' into N8N-5129-N8N-5130-N8N-4442

netroy committed:
Merge remote-tracking branch 'origin/master' into N8N-5129-N8N-5130-N8N-4442

netroy committed:
fix vscode-eslint issues with tsconfig paths

netroy commented:
`vscode-eslint` issues fixed

ivov commented on packages/cli/scripts/build.mjs:
```diff
@@ -54,11 +45,11 @@ function bundleOpenApiSpecs(rootDir = ROOT_DIR, specFileName = SPEC_FILENAME) {
 		.forEach((specPath) => {
 			const distSpecPath = path.resolve(rootDir, 'dist', specPath);
-			const command = `swagger-cli bundle ${specPath} --type yaml --outfile ${distSpecPath}`;
+			const command = `npm run swagger -- bundle src/${specPath} --type yaml --outfile ${distSpecPath}`;
```
Not sure if `/` might cause issues for Win.

netroy replied:
Tested on windows, and it seems to work without any issues.

ivov commented on packages/core/src/ActiveWorkflows.ts:
```diff
@@ -20,8 +20,7 @@ import {
 
-// eslint-disable-next-line import/no-cycle
-import type { IWorkflowData } from '.';
+import type { IWorkflowData } from './Interfaces';
```
Amazing work removing all cycles!

ivov reviewed (commented):
Seeing this issue from 1 above:

```
n8n:build: > n8n@0.201.0 build
n8n:build: > rm -rf dist && tsc -p tsconfig.build.json && tsc-alias -p tsconfig.build.json && node scripts/build.mjs
n8n:build:
n8n:build: src/CurlConverterHelper.ts(2,27): error TS1261: Already included file name '/Users/ivov/Development/n8n/packages/cli/src/curlconverter.d.ts' differs from file name '/Users/ivov/Development/n8n/packages/cli/src/CurlConverter.d.ts' only in casing.
n8n:build:   The file is in the program because:
n8n:build:     Imported via 'curlconverter' from file '/Users/ivov/Development/n8n/packages/cli/src/CurlConverterHelper.ts'
n8n:build:     Matched by include pattern 'src/**/*.ts' in 'tsconfig.build.json'
n8n:build: npm ERR! Lifecycle script `build` failed with error:
n8n:build: npm ERR! Error: command failed
n8n:build: npm ERR!   in workspace: n8n@0.201.0
n8n:build: npm ERR!   at location: /Users/ivov/Development/n8n/packages/cli
n8n:build: ERROR: command finished with error: command (packages/cli) npm run build exited (1)
command (packages/cli) npm run build exited (1)
```

netroy committed:
rename `curlconverter.d.ts` to avoid issues on case-insensitive filesystems

netroy committed:
Merge remote-tracking branch 'origin/master' into N8N-5129-N8N-5130-N8N-4442

netroy committed:
Merge remote-tracking branch 'origin/master' into N8N-5129-N8N-5130-N8N-4442

valya reviewed (dismissed):
Only thing I'm a little wary on is the DB migrations being formatted. It makes it a bit hard to tell if anything's changed in them but it's not worth blocking on.

netroy commented:
> Only thing I'm a little wary on is the DB migrations being formatted. It makes it a bit hard to tell if anything's changed in them but it's not worth blocking on.

Those changes were done by `prettier` automatically. I'll revert those to make the change more readable.

netroy committed:
Merge remote-tracking branch 'origin/master' into N8N-5129-N8N-5130-N8N-4442

netroy committed:
undo re-formating on migration files

netroy committed:
remove yet another import from '.'

netroy commented:
@valya I removed the reformatting on the migration files. can you please have a look again?

netroy committed:
stop deleting `dist` on every build. this breaks incremental build

janober commented:
Got released with `n8n@0.202.0`

dxh9845 commented:
Sorry to bump on this, but when building the docker custom-image we are getting an error when building the n8n-editor-ui. Specifically, the vite build is failing to resolve references to `n8n-workflow` trouble referencing the n8n

```
n8n-editor-ui:build:  ERROR  [vite:resolve] Failed to resolve entry for package "n8n-workflow". The package may have incorrect main/module/exports specified in its package.json: Failed to resolve entry for package "n8n-workflow". The package may have incorrect main/module/exports specified in its package.json.
n8n-editor-ui:build:
n8n-editor-ui:build:
n8n-editor-ui:build:  ERROR  error during build:
n8n-editor-ui:build: Error: Failed to resolve entry for package "n8n-workflow". The package may have incorrect main/module/exports specified in its package.json: Failed to resolve entry for package "n8n-workflow". The package may have incorrect main/module/exports specified in its package.json.
n8n-editor-ui:build:     at packageEntryFailure (/home/node/node_modules/.pnpm/vite@2.9.5_sass@1.55.0/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:38429:11)
n8n-editor-ui:build:     at resolvePackageEntry (/home/node/node_modules/.pnpm/vite@2.9.5_sass@1.55.0/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:38425:9)
n8n-editor-ui:build:     at tryNodeResolve (/home/node/node_modules/.pnpm/vite@2.9.5_sass@1.55.0/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:38232:20)
n8n-editor-ui:build:     at Object.resolveId (/home/node/node_modules/.pnpm/vite@2.9.5_sass@1.55.0/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:38040:28)
n8n-editor-ui:build:     at /home/node/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/shared/rollup.js:22879:40
n8n-editor-ui:build:     at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

netroy commented:
@dxh9845 can you please try building on this commit [52f38df8672120a35fe2e4cc00db019561291421](https://github.com/n8n-io/n8n/commit/52f38df8672120a35fe2e4cc00db019561291421) ?
